routing/router.go

```diff
@@ -1971,34 +1964,48 @@ func pruneVertexFailure(paySession *paymentSession, errSource *btcec.PublicKey)
 // pruneEdgeFailure will attempts to prune an edge from the current available
 // edges of the target payment session in response to an encountered routing
 // error.
-func pruneEdgeFailure(paySession *paymentSession, badChan uint64) {
+func pruneEdgeFailure(paySession *paymentSession, badChan edgeSet) {
 	// We'll inform mission control of this failure so future attempts
 	// avoid this link temporarily.
 
-	paySession.ReportChannelFailure(badChan)
+	paySession.ReportEdgeSetFailure(badChan)
 }
 
-func getFailedChannelID(route *Route, errSource *btcec.PublicKey) (
-	uint64, error) {
+func getFailedEdgeSet(route *Route, errSource *btcec.PublicKey) (
+	edgeSet, error) {
+
+	errSourceVertex := NewVertex(errSource)
 
 	// As this error indicates that the target channel was unable to carry
 	// this HTLC (for w/e reason), we'll query the index to find the
 	// _outgoing_ channel the source of the error was meant to pass the
 	// HTLC along to.
-	if badChan, ok := route.nextHopChannel(errSource); ok {
-		return badChan.ChannelID, nil
+
+	if nextNode, ok := route.nextHopVertex(errSource); ok {
```

Comment:
Interesting. I don't think I changed the logic for determining the problematic channel. In master it is done similarly:

https://github.com/lightningnetwork/lnd/blob/dcd8190c801c3f878d0f58cb3fe0ebf8b0dd2a7e/routing/router.go#L2006

Will think about this. Actually, if we stop making assumptions completely, if we receive an error from a remote node, the only thing we really know is that it is coming from that node and that it can be related to any of his incoming channels from the prev node or outgoing channels to the next node. So an even broader scope than the set of just the outgoing channels. We cannot even know which channels those are, because they can be private (node chooses private channel instead of the specified one). 

Do we want to assume that nodes return honest failure messages? If we don't want to assume that, we could just as well threat all errors identical. Take their type as just an identifier to keep an administration for statistical purposes in the probability machine.

Reply:
We consider the outgoing channel here as if the node is operating as normal, it arrived via the _incoming_ channel. the only time we consider the incoming channel as the channel being failed is for the destination node.

Reply:
> Do we want to assume that nodes return honest failure messages?

The current logic assumes that to a degree. It seems in order to remove this assumption, we'd only need to perform a sanity check (if the failure contains a channel update, maybe they all should?) to ensure we prune the correct edge set.